packages/api/src/utils/psa.js

```diff
@@ -56,19 +60,185 @@ export const psaStatusesToDBStatuses = (statuses) => {
 }
 
 // Error messages
-// TODO: Refactor errors
-export const ERROR_CODE = 400
-export const ERROR_STATUS = 'INVALID_PIN_DATA'
-export const INVALID_CID = 'Invalid cid'
-export const INVALID_MATCH = 'Match should be a string (i.e. "exact", "iexact", "partial", "ipartial")'
-export const INVALID_META = 'Meta should be an object with string values'
-export const INVALID_NAME = 'Name should be a string'
-export const INVALID_ORIGINS = 'Origins should be an array of strings'
-export const INVALID_REQUEST_ID = 'Request id should be a string containing digits only'
-export const INVALID_REPLACE = 'Existing and replacement CID are the same'
-export const INVALID_STATUS = 'Status should be a list of "queued", "pinning", "pinned", or "failed"'
-export const INVALID_TIMESTAMP = 'Should be a valid timestamp'
-export const INVALID_LIMIT = 'Limit should be a number'
+export const DATA_NOT_FOUND = 'Requested data was not found.'
+export const INVALID_CID = 'The CID provided is invalid.'
+export const INVALID_REQUEST_ID = 'Request id should be a string containing digits only.'
+export const INVALID_REPLACE = 'Existing and replacement CID are the same.'
 export const REQUIRED_CID = 'CID is required'
 export const REQUIRED_REQUEST_ID = 'Request id is required'
-export const UNPERMITTED_MATCH = 'Match should be "exact", "iexact", "partial", or "ipartial"'
+
+export const DEFAULT_PIN_LISTING_LIMIT = 10
+export const MAX_PIN_LISTING_LIMIT = 1000
+
+// Validation schemas
+const listPinsValidator = new Validator({
+  type: 'object',
+  required: [],
+  properties: {
+    name: { type: 'string', maxLength: 255 },
+    after: { type: 'string', format: 'date-time' },
+    before: { type: 'string', format: 'date-time' },
+    cid: { type: 'array', items: { type: 'string' } },
+    limit: { type: 'integer', minimum: 1, maximum: MAX_PIN_LISTING_LIMIT },
+    meta: { type: 'object' },
+    match: {
+      type: 'string',
+      enum: ['exact', 'iexact', 'ipartial', 'partial']
+    },
+    statuses: {
+      type: 'array',
+      items: {
+        type: 'string',
+        enum: ['queued', 'pinning', 'pinned', 'failed']
+      }
+    }
+  }
+})
+
+const postPinValidator = new Validator({
+  type: 'object',
+  required: ['cid'],
+  properties: {
+    cid: { type: 'string' },
+    name: { type: 'string', maxLength: 255 },
+    meta: { type: 'object', keys: { type: 'string' } },
+    origins: { type: 'array', items: { type: 'string' }, maxItems: 20 },
+    requestId: { type: 'string' }
+  }
+})
+
+/**
+ * Helper function to parse and validate payload for POST endpoint.
+ *
+ * @param {*} payload
+ * @returns
+ */
+export function validatePinObject (payload) {
+  /** @type {*} */
+  const opts = {}
+  const {
+    cid,
+    name,
+    meta,
+    origins,
+    requestId
+  } = payload
+
+  if (!cid) {
+    return {
+      error: new PSAErrorRequiredData(REQUIRED_CID),
+      data: undefined
+    }
+  }
+
+  // Validate CID.
+  try {
+    opts.cid = normalizeCid(cid)
```

Comment:
Similarly to my other comment for the listing, sourceCid should not be overriden by the normalised one given sourceCid is used as a source of truth in cluster/db operations.

When we are creating a pin, we should pass both the source cid and the normalised one. If you have a look to createPin function we pass both the normalised pin as first parameter and the pinObject containing the sourceCid.